core/src/main/scala/kafka/log/ProducerIdMapping.scala

```diff
@@ -0,0 +1,366 @@
+
+    val fos = new FileOutputStream(file)
+    try {
+      fos.write(buffer.array, buffer.arrayOffset, buffer.limit)
+    } finally {
+      fos.close()
+    }
+  }
+
+  private def verifyFileName(name: String): Boolean = FilenamePattern.findFirstIn(name).isDefined
+
+  private def offsetFromFile(file: File): Long = {
+    s"${file.getName.replace(s".$FilenameSuffix", "")}".toLong
+  }
+
+  private def formatDirName(topicPartition: TopicPartition): String = {
+    s"$DirnamePrefix-${topicPartition.topic}-${topicPartition.partition}"
+  }
+
+  private def formatFileName(lastOffset: Long): String = {
+    s"$lastOffset.$FilenameSuffix"
+  }
+
+}
+
+/**
+ * Maintains a mapping from identifiers to a triple:
+ *   <sequence number, epoch, offset>
+ *
+ * The sequence number is the last number successfully appended to the partition for the given identifier.
+ * The epoch is used for fencing against zombie writers. The offset is the one of the last successful message
+ * appended to the partition.
+ */
+@nonthreadsafe
+class ProducerIdMapping(val config: LogConfig,
+                        val topicPartition: TopicPartition,
+                        val snapParentDir: File,
+                        val maxPidExpirationMs: Int) extends Logging {
+  import ProducerIdMapping._
+
+  val snapDir: File = new File(snapParentDir, formatDirName(topicPartition))
+  snapDir.mkdir()
+
+  private val pidMap = mutable.Map[Long, PidEntry]()
+  private var lastMapOffset = -1L
+  private var lastSnapOffset = -1L
+
+  /**
+   * Returns the last offset of this map
+   */
+  def mapEndOffset = lastMapOffset
+
+  def activePids: immutable.Map[Long, PidEntry] = pidMap.toMap
+
+  /**
+   * Load a snapshot of the id mapping or return empty maps
+   * in the case the snapshot doesn't exist (first time).
+   */
+  private def loadFromSnapshot(logEndOffset: Long) {
+    pidMap.clear()
+    lastSnapshotFile(logEndOffset) match {
+      case Some(file) =>
+        try {
+          trace("Loading snapshot from file: %s".format(file.toPath))
+          loadSnapshot(file, pidMap)
+          lastSnapOffset = offsetFromFile(file)
+          lastMapOffset = lastSnapOffset
+        } catch {
+          case e: CorruptSnapshotException =>
+            // TODO: Delete the file and try to use the next snapshot
+            throw e
+        }
+
+      case None =>
+        lastSnapOffset = -1
+        lastMapOffset = -1
+        snapDir.mkdir()
+    }
+  }
+
+  def checkForExpiredPids(currentTimeMs: Long) {
+    pidMap.retain { case (pid, lastEntry) =>
+      currentTimeMs - lastEntry.timestamp < maxPidExpirationMs
+    }
+  }
+
+  def truncateAndReload(logEndOffset: Long) = {
+    truncateSnapshotFiles(logEndOffset)
+    loadFromSnapshot(logEndOffset)
+  }
+
+  /**
+   * Update the mapping to the given epoch and sequence number.
+   */
+  def update(pid: Long, pidEntry: PidEntry): Unit = {
+    if (pid >= RecordBatch.NO_PRODUCER_ID) {
```

Comment:
This should either be `>` or `>= 0`. We could also move this check to the caller. Either way, we probably need a test case.